### ISSUE TITLE
rpy2 3.5.4 automatically convert object of R class and lose some information

### ISSUE THREAD

**Issue by Fleyderer** [rpy2/rpy2#944]
**Describe the issue or bug**

I am using rpy2 to add support of Rssa (R package) in Python – pyrssa (It is my package name).

rpy2 version 3.4.2 works great, but after switching to 3.5.2 my module is getting errors.

This is my class:

pyrssa/classes/SSA:
```
r_ssa = rpackages.importr('Rssa')
ssa_get = robjects.r('utils::getFromNamespace("$.ssa", "Rssa")')
ssa_contributions = robjects.r('utils::getFromNamespace("contributions", "Rssa")')


class SSA:

    def __init__(self, ds, L, kind="1d-ssa"):
        self.obj = r_ssa.ssa(ds, L=L, kind=kind)
        self.sigma = ssa_get(self.obj, "sigma")
        self.U = ssa_get(self.obj, "U").T
        self.V = ssa_get(self.obj, "V")

    def contributions(self, idx):
        return ssa_contributions(self.obj, idx)
```

The problem is that some fields of ssa class are hidden under "$.ssa" function inside Rssa package, and after new rpy2 auto conversion this object loses access to hidden fields and when i call `self.sigma = ssa_get(self.obj, "sigma")` it produces error: 

"use of NULL environment is defunct"

**To Reproduce**

If you want to reproduce error, you can simply install module using `pip install pyrssa`

and run this code:

```
import pyrssa as prs
import pandas as pd

AustralianWine = prs.data("AustralianWine")
fort = AustralianWine['Fortified'][:174]
fort.index = pd.date_range(start='1980/01/01', freq='M', periods=len(fort))
s_fort = prs.ssa(fort, L=84, kind="1d-ssa")
r_fort = prs.reconstruct(s_fort, groups={"Trend": 1, "Seasonality": range(2, 12)})

prs.plot(r_fort, x=fort.index, add_original=True, add_residuals=True, method="xyplot", superpose=True)
```

If you have rpy2==3.4.2, there is no bug and all works great
If you have rpy2==3.5.4 it gives error

**Expected behavior**

I want to get access to hidden fields of ssa-class objects, such as "sigma", "U", "V".
Maybe, there is a way to prevent this auto conversion only for rssa objects (i've been searching in documentation and couldn't make my own conversion for objects from this package)

**Error**

Traceback (most recent call last):
  File "D:\Python\pyrssa_check\check.py", line 7, in <module>
    s_fort = prs.ssa(fort, L=84, kind="1d-ssa")
  File "D:\Python\pyrssa_check\venv\lib\site-packages\pyrssa\base.py", line 177, in ssa
    return SSA(ds, L, kind)
  File "D:\Python\pyrssa_check\venv\lib\site-packages\pyrssa\classes\SSA.py", line 14, in __init__
    self.sigma = ssa_get(self.obj, "sigma")
  File "D:\Python\pyrssa_check\venv\lib\site-packages\rpy2\robjects\functions.py", line 203, in __call__
    return (super(SignatureTranslatedFunction, self)
  File "D:\Python\pyrssa_check\venv\lib\site-packages\rpy2\robjects\functions.py", line 126, in __call__
    res = super(Function, self).__call__(*new_args, **new_kwargs)
  File "D:\Python\pyrssa_check\venv\lib\site-packages\rpy2\rinterface_lib\conversion.py", line 45, in _
    cdata = function(*args, **kwargs)
  File "D:\Python\pyrssa_check\venv\lib\site-packages\rpy2\rinterface.py", line 815, in __call__
    raise embedded.RRuntimeError(_rinterface._geterrmessage())
rpy2.rinterface_lib.embedded.RRuntimeError: Error in exists(name, envir = .storage(x), inherits = FALSE) : 
  use of NULL environment is defunct

**Additional context**

I've found adding support of "lm" class from R in rpy2 documentation, but still don't understand how to make coverance for ssa class. Maybe you will find some solution and make it work as it expected

**Comment by Fleyderer** [comment]
Now i see that it is converted to OrdDict and after `robjects.r.str(r_ssa.ssa(ds, L=L, kind=kind))` i have:

```
List of 7
 $ call      :List of 7
  ..$ call      : language (function (x, L = (N + 1)%/%2, neig = NULL, mask = NULL, wmask = NULL,      column.projector = "none", row.projec| __truncated__ ...
  ..$ ecall     : language ssa(x = c(`1980-01-31 00:00:00` = 2585L, `1980-02-29 00:00:00` = 3368L,  `1980-03-31 00:00:00` = 3210L, `1980-04-| __truncated__ ...
  ..$ kind      : chr "1d.ssa"
  ..$ svd.method: chr "eigen"
  ..$ fields    : chr [1:12] "F" "wmask" "fmask" "weights" ...
  ..$ length    : int 174
  ..$ window    : int 84
  ..- attr(*, ".env")=<environment: 0x000001d2123ae990> 
  ..- attr(*, "class")= chr [1:2] "1d.ssa" "ssa"
 $ ecall     :List of 7
  ..$ call      : language ssa(x = c(`1980-01-31 00:00:00` = 2585L, `1980-02-29 00:00:00` = 3368L,  `1980-03-31 00:00:00` = 3210L, `1980-04-| __truncated__ ...
  ..$ ecall     : language ssa(x = c(`1980-01-31 00:00:00` = 2585L, `1980-02-29 00:00:00` = 3368L,  `1980-03-31 00:00:00` = 3210L, `1980-04-| __truncated__ ...
  ..$ kind      : chr "1d.ssa"
  ..$ svd.method: chr "eigen"
  ..$ fields    : chr [1:12] "F" "wmask" "fmask" "weights" ...
  ..$ length    : int 174
  ..$ window    : int 84
  ..- attr(*, ".env")=<environment: 0x000001d21180ecf0> 
  ..- attr(*, "class")= chr [1:2] "1d.ssa" "ssa"
 $ kind      : chr "1d.ssa"
 $ svd.method: chr "eigen"
 $ fields    : chr [1:12] "F" "wmask" "fmask" "weights" ...
 $ length    : int [1(1d)] 174
 $ window    : int [1(1d)] 84
<rpy2.rinterface_lib.sexp.NULLType object at 0x000001D20DBC0680> [RTYPES.NILSXP]
```

when originally in R this code gives:

```
List of 7
 $ call      : language ssa(x = fort, L = 84, kind = "1d-ssa")
 $ ecall     : language ssa(x = c(2585L, 3368L, 3210L, 3111L, 3756L, 4216L, 5225L, 4426L, 3932L, 3816L, 3661L, 3795L, 2285L, 2934L, 2985L| __truncated__ ...
 $ kind      : chr "1d.ssa"
 $ svd.method: chr "eigen"
 $ fields    : chr [1:12] "F" "wmask" "fmask" "weights" ...
 $ length    : int 174
 $ window    : num 84
 - attr(*, ".env")=<environment: 0x00000227be214810> 
 - attr(*, "class")= chr [1:2] "1d.ssa" "ssa"
```

So in 3.5.4 idk but default converter somehow makes two lists inside list and changes its class to OrdDict... I really don't understand this logic but i want to use classes of Rssa package just like in R, without these strange wrong conversions

**Comment by Fleyderer** [comment]
Sorry for making many comments here, but i am checking for working in different versions:

3.5.4 – error above
3.5.3 – same as 3.5.4
3.5.2:

```
File "D:\Python\pyrssa\examples\examples_2_1_1_and_2_1_2.py", line 7, in <module>
    s_fort = prs.ssa(fort, L=84, kind="1d-ssa")
  File "D:\Python\pyrssa\pyrssa\base.py", line 178, in ssa
    return SSA(ds, L, kind)
  File "D:\Python\pyrssa\pyrssa\classes\SSA.py", line 12, in __init__
    self.obj = r_ssa.ssa(ds, L=L, kind=kind)
  File "D:\Python\pyrssa\venv\lib\site-packages\rpy2\robjects\functions.py", line 202, in __call__
    return (super(SignatureTranslatedFunction, self)
  File "D:\Python\pyrssa\venv\lib\site-packages\rpy2\robjects\functions.py", line 126, in __call__
    res = cv.rpy2py(res)
  File "D:\Python\Versions\310\lib\functools.py", line 889, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "D:\Python\pyrssa\venv\lib\site-packages\rpy2\robjects\__init__.py", line 237, in _rpy2py_listsexp
    return cls(obj)
  File "D:\Python\Versions\310\lib\functools.py", line 889, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "D:\Python\pyrssa\venv\lib\site-packages\rpy2\robjects\numpy2ri.py", line 207, in rpy2py_sexp
    res = ro.default_converter.rpy2py(obj)
  File "D:\Python\Versions\310\lib\functools.py", line 889, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "D:\Python\pyrssa\venv\lib\site-packages\rpy2\robjects\__init__.py", line 237, in _rpy2py_listsexp
    return cls(obj)
  File "D:\Python\Versions\310\lib\functools.py", line 889, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "D:\Python\pyrssa\venv\lib\site-packages\rpy2\robjects\numpy2ri.py", line 207, in rpy2py_sexp
    res = ro.default_converter.rpy2py(obj)
```
^ and this is endless error

3.5.1: same as 3.5.2
3.5.0: same as 3.5.2
3.4.5: all works great.

Maybe this information will be helpful

**Comment by lgautier** [comment]
```
rpy2.rinterface_lib.embedded.RRuntimeError: Error in exists(name, envir = .storage(x), inherits = FALSE) :
use of NULL environment is defunct
```

The error message is an `RRuntimeError` exception.. This means that the error originates from R when calling it.
The call triggering the exception.

```python
       self.sigma = ssa_get(self.obj, "sigma")
```
R is called with `ssa_get`:

```python
ssa_get = robjects.r('utils::getFromNamespace("$.ssa", "Rssa")')

`self.obj` is defined on the line before `self.sigma`:
```python
        self.obj = r_ssa.ssa(ds, L=L, kind=kind)
```

What you report for `robjects.r.str(r_ssa.ssa(ds, L=L, kind=kind))` seems to match pretty much what R does. Where is the `OrdDict`?  Is this what `type(r_ssa.ssa(ds, L=L, kind=kind))` returns?

**Comment by Fleyderer** [comment]
I am sorry, but I see that `robjects.r.str(r_ssa.ssa(ds, L=L, kind=kind))` creates a different object and you can see it in my comment above (at least the object is duplicated and a list within a list is created). In versions 3.4.* object `self.obj = r_ssa.ssa(ds, L=L, kind=kind)` had type `ListVector` and its behaviour was OK, but now it is `OrdDict` and i have no access to hidden functions and values of classes from Rssa package. 

Yes, it is what `type(r_ssa.ssa(ds, L=L, kind=kind))` returns in versions 3.5.*

**Comment by lgautier** [comment]
There is likely something missing from your example. `rpy2` does not convert lists to `OrdDict` by default:

```python
from rpy2.robjects.packages import importr
base = importr('base')

print(repr(base.list()))
```
```
<rpy2.robjects.vectors.ListVector object at 0x7fde6723a300> [RTYPES.VECSXP]
R classes: ('list',)
[]
```

Also, I see that you are using Windows. Unless using WLS support is going to very limited.

**Comment by Fleyderer** [comment]
I didn't ever see this type before in Python/R,  I don't know, why it happened and that's why i ask here (:

Problem is that all worked fine in previous versions of rpy2, but now it has been broken, so i dont understand where is my mistake in using this package. To see full example you can just install package pyrssa using pip and change rpy2 version to newest and reproduce bug. 

btw, maybe there is another solution which will help to prevent rpy2 from this conversion.

**Comment by lgautier** [comment]
It is hard to tell without a complete picture of what is happening in your Python code (without considering what might be specific to Windows). Try:

```python
class SSA:

    def __init__(self, ds, L, kind="1d-ssa"):
        self.obj = r_ssa.ssa(ds, L=L, kind=kind)
        # Conversion rules are likely not the orginal default one at this point.
        # What does the following print statement show?
        print(rpy2.robjects.conversion.get_conversion())
        # **IF** the default conversion was not modified it is possible to enforce its use.
        with rpy2.robjects.conversion.localconverter(rpy2.robjects.default_converter):
            self.sigma = ssa_get(self.obj, "sigma")
        self.U = ssa_get(self.obj, "U").T
        self.V = ssa_get(self.obj, "V")

    def contributions(self, idx):
        return ssa_contributions(self.obj, idx)
```

**Comment by Fleyderer** [comment]
Thanks for trying to help, I've added this code and checked behavior.

First of all, print statement for conversion:

`<class 'rpy2.robjects.conversion.Converter'>
py2rpy
- builtins.object
- rpy2.rinterface._MissingArgType
- builtins.bool
- builtins.int
- builtins.float
- builtins.bytes
- builtins.str
- rpy2.robjects.robject.RObject
- rpy2.rinterface_lib.sexp.Sexp
- array.array
- builtins.list
- rpy2.rlike.container.TaggedList
- builtins.complex
- builtins.function
- builtins.NoneType
- builtins.range
- builtins.dict
- pyrssa.classes.SSA.SSA
- numpy.ndarray
- numpy.integer
- numpy.floating
- pandas.core.frame.DataFrame
- pandas.core.indexes.base.Index
- pandas.core.series.Series
---
rpy2py
- builtins.object
- rpy2.robjects.robject.RObject
- rpy2.rinterface.IntSexpVector
  - factor (in rpy2.robjects.pandas2ri)
- rpy2.rinterface.FloatSexpVector
  - Date (in rpy2.robjects.vectors)
  - POSIXct (in rpy2.robjects.vectors)
- rpy2.rinterface.ComplexSexpVector
- rpy2.rinterface.BoolSexpVector
- rpy2.rinterface_lib.sexp.StrSexpVector
- rpy2.rinterface.ByteSexpVector
- rpy2.rinterface.PairlistSexpVector
- rpy2.rinterface.LangSexpVector
- rpy2.rinterface.SexpClosure
- rpy2.rinterface_lib.sexp.SexpEnvironment
- rpy2.rinterface.ListSexpVector
  - data.frame (in rpy2.robjects.pandas2ri)
- rpy2.rinterface.SexpS4
- rpy2.rinterface.SexpExtPtr
- rpy2.rinterface_lib.sexp.NULLType
- rpy2.rinterface_lib.sexp.Sexp
- rpy2.rinterface_lib.sexp.SexpVector
- rpy2.robjects.vectors.POSIXct
- rpy2.robjects.vectors.DataFrame
`

I have my own conversion rules for some objects, but no any for SSA from Rssa package. To check this i moved `self.obj = ...` under line with default converter. Of course i got warning with pandas Series type, because ds object is pandas.Series, so i did this:

`
from rpy2 import robjects
from rpy2.robjects.pandas2ri import converter as pandas_converter

class SSA:

    def __init__(self, ds, L, kind="1d-ssa"):
        self.L = L
        self.kind = kind
        with robjects.conversion.localconverter(robjects.default_converter + pandas_converter):
            self.obj = r_ssa.ssa(ds, L=L, kind=kind)
            self.sigma = ssa_get(self.obj, "sigma")
            self.U = ssa_get(self.obj, "U").T
            self.V = ssa_get(self.obj, "V")`

And line with `self.obj =` worked perfectly, but on the next line i got warning:

`Conversion 'py2rpy' not defined for objects of type '<class 'rpy2.rlike.container.OrdDict'>'`

So i see that now, in new versions of rpy2, SSA class becomes OrdDict, while using defalut_converter + pandas_converter. I dont think that problem is in pandas, because this is not about its types, so i see that default converter now creates OrdDict from SSA object and lose information that i need.

I think that problem not in Windows too, because in older versions all conversions works perfectly.

**Comment by lgautier** [comment]
I don't think that the conversion rule from Python to R (`py2rpy`) for your class SSA will convert to OrdDict. When faced with the task of doing that conversion, it should fail with the following by default:
```
NotImplementedError: Conversion 'py2rpy' not defined for objects of type 'pyrssa.classes.SSA.SSA'>'
```

**Comment by michalk8** [comment]
> And line with self.obj = worked perfectly, but on the next line i got warning:
> Conversion 'py2rpy' not defined for objects of type '<class 'rpy2.rlike.container.OrdDict'>'
> So i see that now, in new versions of rpy2, SSA class becomes OrdDict, while using defalut_converter + pandas_converter. 

I encountered the same issue in https://github.com/theislab/cellrank/pull/1008, switching the converter order to
```python
with robjects.conversion.localconverter(pandas_converter + robjects.default_converter):
    ...
```
seems to have solved this issue.

**Comment by Fleyderer** [comment]
Thanks, michalk8!

This solution really works. But there is another problem with numpy conversions too: when you change order of converters, of course you lose conversions from R vectors to numpy arrays. Possible solution is always use "np.asarray(any_r_vector)", but it is so annoying.

I think, default converter should be changed so that the converters order does not affect the result, as it was in rpy2<=3.4.5

Maybe you have comments and solutions for this problem? I think, my problem does not reveal the whole picture, and there are additional pitfalls with converters

Unfortunatelly, for every rpy2=3.5.* versions i have to restrict its usage in my package or have crutches for some conversions.

**Comment by lgautier** [comment]
Stacking conversion rules are order-dependent since it is possible to stack them.

Consider the following stack `ruleset_a + ruleset_b + ruleset_c`  where b and c define conversion rules for `pandas.DataFrame`.

It will look like this as stack. The conversion rules for `pandas.DataFrame` will be ones defined in `ruleset_c` :

```
ruleset_c - pandas.DataFrame
ruleset_b - pandas.DataFrame
ruleset_a - 
```

If changing the order to this, the rules for `pandas.DataFrame` will be in the ones defined in `ruleset_b`

```
ruleset_a
ruleset_b - pandas.DataFrame
ruleset_c - pandas.DataFrame
```

**Comment by Fleyderer** [comment]
I understand that these rules are order-dependent, this conclusion is not very difficult to make.

Problem is that now these rules make conversions for R functions results, and in every version of package all works nice, but in 3.5.* it is broken. Why pandas or numpy conversion rules interfere where it should not be done? They were made for dataframes and arrays, not for random functions.

So what should I do? Just wait for every new problem and create crutches for every new conversion, which will break another piece of functionality?

**Comment by krassowski** [comment]
Somewhat annoyingly I get `Conversion 'py2rpy' not defined for objects of type '<class 'rpy2.rlike.container.OrdDict'>'` a after using `%R -o my_list` when trying to feed `my_list` to R functions. Could it be related?

Minimal reproducible example:

```
%%R -o my_list
my_list = list(a=1, b=2)
```

Followed by:
```python
from rpy2.robjects.packages import importr
base = importr('base')
base.names(my_list)
```

Tested with rpy2 version 3.5.11 and 3.5.13

**Comment by krassowski** [comment]
Another problematic case, when running:

```
%%R -o result
result = list(
    model=lm(mpg ~ am, mtcars)
)
```

not only `result` becomes an `OrdDict` (this might be fine), but `result['model']` is also converted to an `OrdDict` (this is bad - I can no longer call methods which expect `lm` object)

**Comment by lgautier** [comment]
The issue with `OrdDict` should be fixed with PR #1056. Note that this is with `numpy2ri` conversion rules. `OrdDict` is not not handled but the default conversion rules at the moment.

Also note that round-tripping is not guaranteed by conversion rules.  If the purpose of an object returned by an R block of code to pass it to an other block of R code. It is better to keep the rpy2 object a proxy to the R objects. In @krassowski 's example this would be:

```python
%%R -o result -c robjects.default_converter
result = list(
    model=lm(mpg ~ am, mtcars)
)
```

**Comment by lgautier** [comment]
Should be fixed. Reopen if not the case.
